pkg/importer/http_data_source.go

```diff
@@ -0,0 +1,271 @@
+	httpReader io.ReadCloser
+	ctx        context.Context
+	cancel     context.CancelFunc
+	// content type expected by the to live on the endpoint.
+	contentType cdiv1.DataVolumeContentType
+	// stack of readers
+	readers *FormatReaders
+	// endpoint the http endpoint to retrieve the data from.
+	endpoint *url.URL
+	// url the url to report to the caller of getURL, could be the endpoint, or a file in scratch space.
+	url *url.URL
+	// true if we are using a custom CA (and thus have to use scratch storage)
+	customCA bool
+}
+
+// NewHTTPDataSource creates a new instance of the http data provider.
+func NewHTTPDataSource(endpoint, accessKey, secKey, certDir string, contentType cdiv1.DataVolumeContentType) (*HTTPDataSource, error) {
+	ep, err := ParseEndpoint(endpoint)
+	if err != nil {
+		return nil, errors.Wrapf(err, fmt.Sprintf("unable to parse endpoint %q", endpoint))
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	httpReader, err := createHTTPReader(ctx, ep, accessKey, secKey, certDir)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	if accessKey != "" && secKey != "" {
+		ep.User = url.UserPassword(accessKey, secKey)
+	}
+	dataProvider := &HTTPDataSource{
+		ctx:         ctx,
+		cancel:      cancel,
+		httpReader:  httpReader,
+		contentType: contentType,
+		endpoint:    ep,
+		customCA:    certDir != "",
+	}
+	// We know this is a counting reader, so no need to check.
+	countingReader := httpReader.(*util.CountingReader)
+	go dataProvider.pollProgress(countingReader, 10*time.Minute, time.Second)
+	return dataProvider, nil
+}
+
+// Info is called to get initial information about the data.
+func (dp *HTTPDataSource) Info() (ProcessingPhase, error) {
+	var err error
+	dp.readers, err = NewFormatReaders(dp.httpReader, dp.contentType)
+	if err != nil {
+		klog.Errorf("Error creating readers: %v", err)
+		return Error, err
+	}
+	// The readers now contain all the information needed to determine if we can stream directly or if we need scratch space to download
+	// the file to, before converting.
+	if dp.readers.Qemu && !dp.readers.Archived && !dp.customCA {
+		// We can pass straight to conversion from the endpoint. No scratch required.
+		dp.url = dp.endpoint
+		return Convert, nil
+	}
+	if dp.contentType == cdiv1.DataVolumeArchive {
+		return TransferTarget, nil
+	}
+	return Transfer, nil
+}
+
+// Transfer is called to transfer the data from the source to a scratch location.
+func (dp *HTTPDataSource) Transfer(path string) (ProcessingPhase, error) {
+	if dp.contentType == cdiv1.DataVolumeKubeVirt {
+		if util.GetAvailableSpace(path) <= int64(0) {
+			//Need scratch space but none provided.
+			return Error, ErrRequiresScratchSpace
+		}
+		file := filepath.Join(path, tempFile)
+		err := StreamDataToFile(dp.readers.TopReader(), file)
+		if err != nil {
+			return Error, err
+		}
+		// If we successfully wrote to the file, then the parse will succeed.
+		dp.url, _ = url.Parse(file)
+		return Process, nil
```

Comment:
No that should work, the format readers are a stack each reader will extract one format and pass the result to the next reader. so for imagefile.qcow2.tar.gz it will create a tar and a gz format reader, which will result in the qcow2 file being written to scratch space, then coverted to raw during the convert phase.

Reply:
Do we have a unit test that checks with case?

Reply:
That is not a unit test instance, we have a functional test that checks though.